### DISCUSSION TITLE
Watchdog fail error

### DISCUSSION THREAD

**Discussion by jmemon16** [bitcraze/discussions#1703] [Q&A] [answered]
Hi,

I am facing difficulty in understanding why my code gives a watchdog fail error and then led 1 blinks with five pulses. I am trying peer to peer communication and using usec_time.h to measure round trip time. 

> What am I missing here?
> I have faced this issue in my other apps as well. But if I disable some instructions and keep code simple then it goes away for example I remove peer2peer communication code (i.e. p2pRegisterCB). 
> Are there any guidelines to follow what we can into app and what cannot?
> How is watchdog configured? Does it have conflict with timer in usec_time?

Below is mode code in app(which is what I am playing around mostly). and console output. 

I will appreciate if you have any tips or tutorial to follow for better understanding on debugging the code crazyflie ecosystem and pin pointing the issue. Thanks

_____________________________________________________________
```

#include <string.h>
#include <stdint.h>
#include <stdbool.h>
#include <stdio.h>


#include "usec_time.h"


#include "app.h"

#include "FreeRTOS.h"
#include "task.h"

#include "radiolink.h"
#include "configblock.h"

//#define DEBUG_MODULE "P2P"
#include "debug.h"

#define MESSAGE1 "tic"
#define MESSAGE2 "toc"
#define MESSAGE_LENGHT 3



static uint8_t my_check_flag_1= 0;
static uint8_t my_check_flag_2= 0;
uint64_t start_time, current_time, log_time; // variables to keep track of time


void p2pcallbackHandler(P2PPacket *p)
{
  // Parse the data from the other crazyflie and print it
  uint8_t other_id = p->data[0];
  //static char msg[MESSAGE_LENGHT + 1];
  //memcpy(&msg, &p->data[1], sizeof(char)*MESSAGE_LENGHT);
  //msg[MESSAGE_LENGHT] = 0;
  //uint8_t rssi = p->rssi;

  if(other_id == 1) {
    //what should drone 2 do?
    DEBUG_PRINT("Tic-Received \n");
    my_check_flag_2= 1;   // send toc
  }
  else{
    //what should drone 1 do?
    DEBUG_PRINT("Toc-Received \n");
    my_check_flag_1= 1;  // send tic
  }

  
}

void appMain()
{
  DEBUG_PRINT("Waiting for activation ...\n");
  vTaskDelay(M2T(3000));
  usecTimerInit();
  
  start_time = usecTimestamp(); 
  DEBUG_PRINT("Time = %llu \n",start_time);
 
  // Initialize the p2p packet 
  static P2PPacket p_reply;
  p_reply.port=0x00;
  uint64_t address = configblockGetRadioAddress();
  uint8_t my_id =(uint8_t)((address) & 0x00000000ff);

  p_reply.data[0]=my_id;

  //Put a string in the payload
  char *str= MESSAGE1;
  memcpy(&p_reply.data[1], str, sizeof(char)*MESSAGE_LENGHT);

  // Set the size, which is the amount of bytes the payload with ID and the string 
  p_reply.size=sizeof(char)*MESSAGE_LENGHT+1;
     

  // Register the callback function so that the CF can receive packets as well.
  //p2pRegisterCB(p2pcallbackHandler);

  while(1) {
    

    if (my_id == 1) {
      switch (my_check_flag_1){
        case 0: // message not sent, send message
          str= MESSAGE1;
          radiolinkSendP2PPacketBroadcast(&p_reply);
          log_time = usecTimestamp();
          DEBUG_PRINT("Sent at: %llu \n",log_time);
          my_check_flag_1 = 2; //waiting for reply
          break;
        case 1: // confirmation of round trip
          current_time = usecTimestamp();
          DEBUG_PRINT("Round trip : %llu \n",(current_time -log_time));
          my_check_flag_1 = 0;  // go back to inital setup after 2 sec.
          vTaskDelay(M2T(2000));
          break;
        case 2: // wait for 5 sec, if no response, send again
          current_time = usecTimestamp();
          if (current_time-log_time>5000000){
            my_check_flag_1 = 0;  // go back to inital setup
            DEBUG_PRINT("Reset exp \n");

          }
          
          break;
        default:
          break;
      }
    }
  
     /*else {
      if (my_check_flag_2 ==1){
        str= MESSAGE2;
        memcpy(&p_reply.data[1], str, sizeof(char)*MESSAGE_LENGHT);
        DEBUG_PRINT("CF %d send %s\n", my_id, str);
        //radiolinkSendP2PPacketBroadcast(&p_reply);
        my_check_flag_2 = 0;
      }
      // do nothing
    }*/

    
  }
}

```

Console output
________________________________________________________________
SYS: Crazyflie 2.1 is up and running!
SYS: Build 22:59fa2f6c6837 (2024.10.2 +22) MODIFIED
SYS: I am 0x313232313232470F0040002B and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 1 deck(s) found
DECK_CORE: Calling INIT on driver bcFlow2 for deck 0
ZR2: Z-down sensor [OK]
PMW: Motion chip id: 0x49:0xB6
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2024.10 (CF21)
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
DECK_CORE: Deck 0 test [OK].
SYS: The system resumed after watchdog timeout [WARNING]
SYS: watchdogNormalStart [FAIL]

**Comment by gemenerik** [comment]
Are you sure there are no cases where the code may loop without sleeping? This could cause problems

**Comment by jmemon16** [comment]
Hi Rik,
Is it necessary to sleep in app? Can we not have indefinite loop in main app?

I think you have pointed out correctly. I noticed that I need to have atleast 1 millisecond delay in a while loop to remove this error.

I wonder how do we avoid that? I think during sleep watchdog is fed. is this right? Is there any instruction to feed watchdog in main app without adding delay of 1 millisecond?

For delay I am using vtaskdelay.

Junaid

On 20 Dec 2024, at 11:24 am, Rik ***@***.***> wrote:

﻿

Are you sure there are no cases where the code may indefinitely loop without sleeping? This could cause problems

**Reply by ToveRumar** [reply, marked as answer]
Hi @jmemon16! 

The watchdog is being fed from the idle task and it has the lowest priority of all tasks. So if your main task  is loading the CPU 100% the idle task will never be able to run and feed the watchdog. 

I would in  general not advice to load the CPU 100%, since this might cause unexpected behaviors, unless you are very knowledgeable on all the priorities of all running tasks.

You can feed the watchdog from your main task by calling watchdogReset(). 
See `void vApplicationIdleHook( void )` in src/modules/system.c to see how it is done in the idle task. 

I hope this answered your question!

**Reply by jmemon16** [reply]
Hi ToveRumar,
Thank you for clarifying that. Where can I see the list of tasks and their priority level?

**Reply by ToveRumar** [reply]
Hi @jmemon16 !
Im happy I could help!

Unfortunately I do not know of any simple function to do this. Maybe there is some FreeRTOS magic that could be applied to get this information. 

However there should be a handful number of tasks, so looking through the src  to find all threads created using xTaskCreate() should not be too big of a job. The Idle task is not created explicitly but it is always running in the background, so this should also be added to the list.
